common/base/tests/Test_UtilsThreadsafeContainers.cpp

```diff
@@ -0,0 +1,237 @@
+    std::vector<std::int32_t> occurrences_per_element(number_of_elements, 0);
+
+    std::vector<std::thread> producer_threads;
+    std::vector<std::thread> consumer_threads;
+
+    for (std::size_t producer_id = 0; producer_id < number_of_producers; ++producer_id)
+    {
+        producer_threads.push_back(std::thread([&producer_consumer, number_of_producers, producer_id, producers_sleep_after, producers_sleep_for_ms, number_of_elements_per_producer]() {
+            const std::size_t producer_offset = producer_id * number_of_elements_per_producer;
+            for (std::size_t i = 0; i < producers_sleep_after; ++i)
+            {
+                producer_consumer.add_new_element(producer_offset + i);
+            }
+            std::this_thread::sleep_for(std::chrono::milliseconds(producers_sleep_for_ms));
+            for (std::size_t i = producers_sleep_after; i < number_of_elements_per_producer; ++i)
+            {
+                producer_consumer.add_new_element(producer_offset + i);
+            }
+            if (number_of_producers == 1)
+            {
+                // if there is only one producer there is no need to wait for other (nonexistent) producers to finish writing
+                producer_consumer.signal_pushed_last_element();
+            }
+        }));
+    }
+
+    for (std::size_t consumer_id = 0; consumer_id < number_of_consumers; ++consumer_id)
+    {
+        consumer_threads.push_back(std::thread([&producer_consumer, &occurrences_per_element, &occurrences_per_element_mutex]() {
+            while (true)
+            {
+                cga_optional_t<std::size_t> val = producer_consumer.get_next_element();
+                if (!val) // reached the end
+                {
+                    break;
+                }
+                else
+                {
+                    std::lock_guard<std::mutex> lg(occurrences_per_element_mutex); // if everything goes well every element will be written to exactly one, but if there is a bug several theads could access the same element
+                    occurrences_per_element[val.value()]++;
+                }
+            }
+        }));
+    }
+
+    for (std::thread& producer_thread : producer_threads)
+    {
+        producer_thread.join();
+    }
+    if (number_of_producers > 1)
+    {
+        producer_consumer.signal_pushed_last_element();
+    }
+    for (std::thread& consumer_thread : consumer_threads)
+    {
+        consumer_thread.join();
+    }
+
+    ASSERT_TRUE(std::all_of(std::begin(occurrences_per_element),
+                            std::end(occurrences_per_element),
+                            [](const std::size_t val) {
+                                return val == 1;
+                            }));
+}
+
+TEST(TestUtilsThreadsafeContainers, test_threadsafe_producer_consumer_single_producer_single_consumer)
```

Comment:
Google test provides a frameworks for parametrized tests with `TEST_P` which can run the same test harness code with different inputs. You only have 2 parametrized versions here, but something to consider if you want to increase the coverage with little effort.

http://sandordargo.com/blog/2019/04/24/parameterized-testing-with-gtest

Reply:
Good point, I'll look into it

Reply:
I looked into it and I'd still stick with my solution, if you are ok with it.
As there are only two tests I think this would actually only make the code more complicated.

Reply:
fine by me.

Reply:
Fair enough. Just wanted to bring this up in case you wanted to expand the scope of this test. Resolving.